apps/src/code-studio/pd/foorm/editor/library/FoormLibrarySaveBar.jsx

```diff
@@ -0,0 +1,482 @@
+        this.setState({
+          showNewLibraryQuestionSave: false
+        });
+      });
+  };
+
+  handleSaveSuccess(libraryQuestion) {
+    this.setState({
+      isSaving: false
+    });
+    this.props.setLastSaved(Date.now());
+    const updatedQuestion = JSON.parse(libraryQuestion.question);
+    // reset code mirror with returned questions (may have added question name)
+    this.props.resetCodeMirror(updatedQuestion);
+    // update store with library question data.
+    this.props.addAvailableLibraryQuestion({
+      id: libraryQuestion['id'],
+      name: libraryQuestion['question_name'],
+      type: JSON.parse(libraryQuestion.question)['type']
+    });
+    this.props.setLibraryQuestionData({
+      name: libraryQuestion.question_name,
+      id: libraryQuestion.id,
+      question: updatedQuestion
+    });
+    this.props.setLastSavedLibraryQuestionQuestions(updatedQuestion);
+  }
+
+  handleSaveError(result) {
+    this.setState({
+      isSaving: false
+    });
+
+    this.props.setSaveError(
+      (result.responseJSON && result.responseJSON.questions) ||
+        result.responseText ||
+        'Unknown error.'
+    );
+  }
+
+  isNameErrorShown() {
+    // True if library question name has been entered and is invalid,
+    // or library name has been entered and is invalid.
+    return (
+      (this.state.libraryQuestionName &&
+        !this.isNameValid(this.state.libraryQuestionName)) ||
+      (this.state.libraryName && !this.isNameValid(this.state.libraryName))
+    );
+  }
+
+  isSaveNewLibraryQuestionDisabled() {
+    // disable if invalid name for library question
+    if (this.isNameErrorShown()) {
+      return true;
+    }
+
+    // for existing library, library question name is all that is required
+    if (this.props.libraryId && !this.state.libraryQuestionName) {
+      return true;
+    }
+
+    // for new library, need library name, category, and a library question name
+    if (
+      !this.props.libraryId &&
+      !(
+        this.state.libraryName &&
+        this.state.libraryCategory &&
+        this.state.libraryQuestionName
+      )
+    ) {
+      return true;
+    }
+
+    return false;
+  }
+
+  renderPublishedSaveWarning(forms) {
+    let formBullets = forms.map((form, i) => <li key={i}>{form}</li>);
+
+    return (
+      <div>
+        <span style={styles.warning}>Warning: </span>This question appears in
+        one or more published surveys, listed below:
+        <ul>{formBullets}</ul>
+        Please only make safe edits as described in the{' '}
+        <a
+          href="https://github.com/code-dot-org/code-dot-org/wiki/%5BLevelbuilder%5d-Foorm-Editor:-Editing-a-Form#safe-edits-to-published-forms"
+          target="_blank"
+          rel="noopener noreferrer"
+        >
+          How To
+        </a>
+        .
+        <br />
+        <br />
+        Are you sure you want to save your changes?
+      </div>
+    );
+  }
+
+  renderNewLibraryQuestionSaveModal = () => {
```

Comment:
There was a (totally reasonable) suggestion when folks first looked at this PR to move this into its own component. I looked at doing that, but found that the state/methods used in this modal are all tangled up with the state/methods used elsewhere in this file.

As it's not a super trivial refactor (eg, move some code into a separate file), and this follows the same pattern in the [existing FoormFormSaveBar component](https://github.com/code-dot-org/code-dot-org/blob/staging/apps/src/code-studio/pd/foorm/editor/form/FoormFormSaveBar.jsx#L260), I'm inclined to do this refactor work as a follow-up if necessary.